### PR TITLE
Add DNS-over-HTTPS P3A metrics

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/31965

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

0. Ensure a DNS server that supports DoH (such as Cloudflare: 1.1.1.1) is configured in your OS network settings. Start with fresh profile.
1. Ensure that the secure setting metric has a value of 1.
2. Make a random query on your search engine, click on a bunch unique links.
3. On the local state page, search for `upgraded_dns_requests` and `total_dns_requests`. Note the values. `upgraded_dns_requests` should have a positive value, if your OS settings are correct.
4. The percentage range indicated by the value of the auto-secure requests metric should encompass the ratio of `upgraded_dns_requests` to `total_dns_requests` (i.e. if the upgraded count is 10, and the total count is 15, the ratio would be 66% and the correct metric answer would be 2)
5. Disable the DoH supporting server in your OS settings (let DHCP do it's thing)
6. Repeat step 2.
7. Ensure that the `upgraded_dns_requests` count does not grow as you visit websites. The ratio of upgraded-to-total requests should become lower as you visit more domains. The auto-secure requests metric value should also change accordingly.
8. Change your browser DoH setting to "With Cloudflare". `Brave.DNS.SecureSetting` should be 2, and `Brave.DNS.AutoSecureRequests` should disappear.
9. Turn off your browser DoH setting. Ensure both metrics do not exist.

### CODE REVIEW

File: browser/misc_metrics/doh_metrics_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace misc_metrics {
+
+class DohMetricsTest : public testing::Test {
+ public:
+  DohMetricsTest() = default;
+
+ protected:
+  void SetUp() override {
+    auto* registry = local_state_.registry();
```
[the remark this answers is not in the document]
Comment (fmarier) [reply browser/misc_metrics/doh_metrics_unittest.cc:25]: @thypon Looks like our regexp is not tight enough here. This is a local variable in an inline function, not a class field.

[the remark this answers is not in the document]
Comment (thypon) [reply browser/misc_metrics/doh_metrics_unittest.cc:25]: Indeed, I think we can solve that with a pattern-not-inside for inline methods

File: test/BUILD.gn [hunk 2]
```diff
@@ -98,6 +98,7 @@ test("brave_unit_tests") {
     "//brave/browser/brave_stats/brave_stats_updater_unittest.cc",
     "//brave/browser/browsing_data/brave_browsing_data_remover_delegate_unittest.cc",
     "//brave/browser/download/brave_download_item_model_unittest.cc",
+    "//brave/browser/misc_metrics/doh_metrics_unittest.cc",
```
Comment (iefremov) [inline test/BUILD.gn:101]: i probably missed this during previous reviews, but we shouldn't populate this section, rather tests should go to their own  BUILD.gn in corresponding directories. Can you please do a follow-up for metrics and time_period storage?

File: browser/misc_metrics/doh_metrics.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+class PrefRegistrySimple;
+class PrefService;
+
+namespace misc_metrics {
+
+extern const char kAutoSecureRequestsHistogramName[];
+extern const char kSecureDnsSettingHistogramName[];
+
+// DNS-over-HTTPS metrics.
+class DohMetrics : public base::SupportsWeakPtr<DohMetrics> {
```
[the remark this answers is not in the document]
Comment (iefremov) [reply browser/misc_metrics/doh_metrics.h:26]: also would be useful to learn how it gathers metrics without looking at the code

File: browser/misc_metrics/doh_metrics.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+DohMetrics::DohMetrics(PrefService* local_state)
+    : total_request_storage_(local_state, kMiscMetricsTotalDnsRequestStorage),
+      upgraded_request_storage_(local_state,
+                                kMiscMetricsUpgradedDnsRequestStorage),
+      local_state_(local_state) {
+  pref_change_registrar_.Init(local_state_);
+  pref_change_registrar_.Add(
+      prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&DohMetrics::HandleDnsOverHttpsMode, AsWeakPtr()));
```
Comment (iefremov) [inline browser/misc_metrics/doh_metrics.cc:43]: base::Unretained is safe here, this is a common pattern

File: test/BUILD.gn (hunk 2 above)
Comment (DJAndries) [reply test/BUILD.gn]: fixed, will make follow-up PRs for other components

**Review by iefremov — APPROVED** [review]

**Comment by GeetaSarvadnya** [comment]
Verification **INPRGORESS** on 

```
Brave | 1.61.20 Chromium: 118.0.5993.70 (Official Build) nightly (64-bit)
-- | --
Revision | ade3aa32a9f6a4eb7d133a575d6683f611006ac6
OS | Windows 10 Version 22H2 (Build 19045.3570)
```

<details>
<summary>Brave.DNS.SecureSetting = 1</summary>

</details>

<details>
<summary>Brave.DNS.SecureSetting = 2</summary>

</details>

-----

<details>
<summary>Brave.DNS.AutoSecureRequests = 0</summary>

</details>

<details>
<summary>Brave.DNS.AutoSecureRequests = 1</summary>

</details>

<details>
<summary>Brave.DNS.AutoSecureRequests = 2</summary>

</details>

<details>
<summary>Brave.DNS.AutoSecureRequests = 3</summary>

</details>

**Comment by GeetaSarvadnya** [comment]
@DJAndries: I have verified the PR on the latest nightly build, the test steps 4 and 7 are failing and rest of the steps are working as per the test plan. 

Could you review the the test plan scenario above https://github.com/brave/brave-core/pull/19479#issuecomment-1760990527? Please let me know if you need more details. Thanks!

cc: @kjozwiak

**Comment by DJAndries** [comment]
> upgrade count is 35.0 and total dns requests count is 43.0 shown in the screenshot

I see you have two screenshots for this step. In the first one, you have an upgraded count of 35 and total of 37. In the second one,  you have an upgraded count of 40 and total of 43. I recommend evaluating the upgraded and total count, as well as the metric value, simultaneously.

> the count value is increased for both metrics without visiting any new websites - not sure why it's so
What is your usual DNS server? Is it just your ISP's server? If so, requests should not be upgraded. The total count should increase however. I recommend waiting around 30 seconds for the browser to idle after switching your DNS settings, and then visiting websites to see if the upgraded count increases.

**Comment by GeetaSarvadnya** [comment]
@DJAndries: Please find the attached gif file below, I am evaluating the `upgraded` and `total count` as well as the metrics value at the same time. I have executed the case two times, I am happy with the case 2 as I am getting the expected result

**Case 1:**

I am getting the upgraded counct value as `9` and total count value as `10` (i.e 9/10*100 = 90% ). I am getting the metrics `Brave.DNS.AutoSecureRequests` response value as `2` I am not sure 90% falls under which repsonse bucket below? there is no reponse bucket for the exact 90 %

```
Brave.DNS.AutoSecureRequests

0. Greater than 0%, less than 5%
1. Greater than 5%, less than 50%
2. Greater than 50%, less than 90%
3. Greater than 90%
```
https://drive.google.com/file/d/1xxWag_TXCt0h-iphItPWiwVY22pDFj_C/view?usp=sharing

**Case 2:**

I am getting the upgraded counct value as `4` and total count value as `4` (i.e 4/4*100 = 100% ). I am getting the metrics `Brave.DNS.AutoSecureRequests` response value as `3` as expected

P3A_1

**Comment by GeetaSarvadnya** [comment]
>What is your usual DNS server? Is it just your ISP's server? If so, requests should not be upgraded. The total count should increase however. I recommend waiting around 30 seconds for the browser to idle after switching your DNS settings, and then visiting websites to see if the upgraded count increases.

@DJAndries My DNS server is ISP. I have waited for 30 sec even more than 1 min after disabling the DoH, before I make a random query on  search engine, when I reload the brave://local-state, the upgraded and total count value is increased.

Please find the gif below
https://drive.google.com/file/d/1lij9L9OQDil7aN2lij0ehOgKGeuYzWUL/view?usp=sharing

**Comment by DJAndries** [comment]
@GeetaSarvadnya 

RE case 1: updated answers to include "less than or equal to". This case looks correct.

RE case 2: Perhaps your ISP DNS supports DoH? I recommend switching to OpenNIC for step 5: https://www.opennic.org/

Also run `ipconfig /all` to ensure that your DNS settings are applied correctly.

**Comment by GeetaSarvadnya** [comment]
Happy with case 1. Thanks for updating the answers.

@DJAndries : RE case 2 - My ISP DNS doesn't support DoH, I checked with ISP. I am getting the same result using https://www.opennic.org/. 
I ran `ipconfig /all` to confim the DNS settigs applied.

**Comment by DJAndries** [comment]
Could you please try restarting the browser after updating the DNS settings?

**Comment by GeetaSarvadnya** [comment]
@DJAndries I have tried restarting the browser as well, Also I have just kept local state file open and closed the opened tabs after disabling the DoH settings from OS but no luck

File: net/dns/secure_dns_counter.cc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  DnsTaskType task_type = static_cast<DnsTaskType>(task_type_int);
+
+  if (!query_types.Has(DnsQueryType::HTTPS)) {
+    // Only report dns tasks that are requesting HTTPS records.
+    return;
+  }
+
+  bool is_upgraded_task = task_type == DnsTaskType::SECURE_DNS;
+
+  counters_lock_.Acquire();
```
Comment (bridiver) [inline net/dns/secure_dns_counter.cc:36]: we should not be locking in a critical code path like this

**Pull request merged**